File: csharp/1358.number-of-substrings-containing-all-three-characters.cs
/*
 * @lc app=leetcode id=1358 lang=csharp
 *
 * [1358] Number of Substrings Containing All Three Characters
 */

using System;

// @lc code=start
public class Solution {
    public int NumberOfSubstrings(string s) {
        int[] indices = new int[] { -1, -1, -1 };
        int n = s.Length, ret = 0;
        for (int i = 0; i < n; ++i) {
            int j = (int)s[i] - (int)'a';
            indices[j] = i;
            ret += Math.Min(indices[(j + 1) % 3], indices[(j + 2) % 3]) + 1;
        }
        return ret;
    }
}
// @lc code=end

File: java/1358.number-of-substrings-containing-all-three-characters.java
/*
 * @lc app=leetcode id=1358 lang=java
 *
 * [1358] Number of Substrings Containing All Three Characters
 */

// @lc code=start
class Solution {
    public int numberOfSubstrings(String s) {
        int n = s.length(), ret = 0;
        int[] prev = new int[] { -1, -1, -1 };
        for (int i = 0; i < n; ++i) {
            int c = s.charAt(i) - 'a';
            prev[c] = i;
            ret += 1 + Math.min(prev[0], Math.min(prev[1], prev[2]));
        }
        return ret;
    }
}
// @lc code=end

